### DISCUSSION TITLE
JK-BMS-PB RS485 CAN BUS

### DISCUSSION THREAD

**Discussion by Sleeper85** [txubelaxu/esphome-jk-bms#8] [General]
Hi,

First of all, thank you for this new component.

I'm integrating your code with my JK-BMS-CAN project.

The person currently testing the code in development has 7x JK-BMS-PB and was already using your code for BMS supervision only (CAN communication between master-BMS and inverter).

This person would like to use my project, the ESP32 would therefore be the master BMS and will communicate with the BMS on the RS485 bus and with the Deye inverter on canbus.

Your code worked before with a classic 38 pin ESP32 board. The current tests are taking place with an Atom S3 Lite and strange things are happening concerning the "online status" of the BMS and probably also of the other sensors (I'm not onsite and I don't have this new model of BMS for testing).

[Watch the video showing the problem](https://drive.google.com/file/d/1YY7K4iDkVaTMGQ7xW1UHR6BweQYSplEf/view?usp=drive_link)

Testing is currently being carried out on a version containing only your component in order to eliminate possible problems coming from my code.

[JK-BMS-PB-CAN solution (how everything is connected together)](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/documents/README/JK-BMS-PB-CAN_solution.md)

[There is an error in your schematic, TX goes to DI and RX to RO !](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/images/RS485_Module_HCMODU0081_Diagram.png)

Where do you think this problem could come from ?

Do you connect the GND on the RS485 bus side?

Best regards and thanks for your help

**Comment by Sleeper85** [comment]
@txubelaxu 

I see you forked my git, maybe you want to try this JK-BMS-PB-CAN version?
I don't have a JK-BMS-PB to test this so I think you are best placed to do it ;-)

Replace the Atom S3 Lite board with your own board.

https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/

**Comment by Sleeper85** [comment]
I found the problem, the code below does not work with an `update_interval` of `1s` or `2s` (7x JK-BMS-PB).
It works with `3s`.

```YAML
jk_rs485_bms:
  - id: bms0
    rs485_address: 0x00
    jk_rs485_sniffer_id: sniffer0
    update_interval: 1s
```

**Comment by txubelaxu** [comment]
Sorry for the delay, @Sleeper85.

I have just tried to compile:

`esphome compile multi-bms_RS485_JK-BMS_PB_series_mode2.yaml`

I have got an "error": 

```
Indexing .pioenvs/smartbms/esp-idf/app_update/libapp_update.a
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_common_loader.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_clock_init.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_flash.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_mem.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_random.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_random_esp32.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_utility.o
Compiling .pioenvs/smartbms/bootloader_support/src/esp_image_format.o
Compiling .pioenvs/smartbms/bootloader_support/src/flash_encrypt.o
Compiling .pioenvs/smartbms/bootloader_support/src/secure_boot.o
Compiling .pioenvs/smartbms/bootloader_support/src/flash_partitions.o
Compiling .pioenvs/smartbms/bootloader_support/src/flash_qio_mode.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_flash_config_esp32.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_efuse.o
Compiling .pioenvs/smartbms/bootloader_support/src/idf/bootloader_sha.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborencoder_close_container_checked.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborencoder.o
In file included from src/esphome/components/jk_rs485_sniffer/jk_rs485_sniffer.h:6,
                 from src/esphome/components/jk_rs485_bms/jk_rs485_bms.h:8,
                 from src/esphome.h:32,
                 from src/main.cpp:3:
src/esphome/components/jk_rs485_sniffer/binary_output.h:18:7: error: redefinition of 'class esphome::output::BinaryOutput'
 class BinaryOutput {
       ^~~~~~~~~~~~
In file included from src/esphome/components/binary/light/binary_light_output.h:4,
                 from src/esphome.h:15,
                 from src/main.cpp:3:
src/esphome/components/output/binary_output.h:18:7: note: previous definition of 'class esphome::output::BinaryOutput'
 class BinaryOutput {
       ^~~~~~~~~~~~
Archiving .pioenvs/smartbms/esp-idf/bootloader_support/libbootloader_support.a
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborerrorstrings.o
Indexing .pioenvs/smartbms/esp-idf/bootloader_support/libbootloader_support.a
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborparser_dup_string.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborparser.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborpretty_stdio.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborpretty.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cbortojson.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborvalidation.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/open_memstream.o
Compiling .pioenvs/smartbms/cmock/CMock/src/cmock.o
```

**Reply by Sleeper85** [reply]
Hi,

With which board?

I encounter the same problem with 'board_esp32-devkit-v1.yaml' regardless of the talk pin I choose. I haven't figured out why yet?

But the code compiles well with ESP32-S3, ESP32-C3.

**Comment by txubelaxu** [comment]
> Hi,
> 
> First of all, thank you for this new component.
> 
> I'm integrating your code with my JK-BMS-CAN project.
> 
> The person currently testing the code in development has 7x JK-BMS-PB and was already using your code for BMS supervision only (CAN communication between master-BMS and inverter).
> 
> This person would like to use my project, the ESP32 would therefore be the master BMS and will communicate with the BMS on the RS485 bus and with the Deye inverter on canbus.
> 
> Your code worked before with a classic 38 pin ESP32 board. The current tests are taking place with an Atom S3 Lite and strange things are happening concerning the "online status" of the BMS and probably also of the other sensors (I'm not onsite and I don't have this new model of BMS for testing).
> 
> [Watch the video showing the problem](https://drive.google.com/file/d/1YY7K4iDkVaTMGQ7xW1UHR6BweQYSplEf/view?usp=drive_link)
> 
> Testing is currently being carried out on a version containing only your component in order to eliminate possible problems coming from my code.
> 
> [JK-BMS-PB-CAN solution (how everything is connected together)](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/documents/README/JK-BMS-PB-CAN_solution.md)
> 
> [There is an error in your schematic, TX goes to DI and RX to RO !](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/images/RS485_Module_HCMODU0081_Diagram.png)
> 
> Where do you think this problem could come from ?
> 
> Do you connect the GND on the RS485 bus side?
> 
> Best regards and thanks for your help

Sorry again for the delay. 
Yes, @Sleeper85, I connect GND as well in a RS485 bus (A, B, GND). 

I would like to help you to integrate my code into your project and testing my two JK-PB BMSs in a "mode 2" (so, none of the bms is acting as master, so ESP will act as master)

I have a problem with my "talk_pin" because I think I am using "binary_output" incorrectly in my code. 

I have no inverter at this moment. I started my system from the battery. lol. But I would like to use your code in my system.

Many thanks!!!

**Reply by Sleeper85** [reply]
Hi,

It's good if you can help to integrate your code into my project.

Concerning RS485 the best boards do not use TALK PIN, can your code work without TALK PIN ?

**Comment by txubelaxu** [comment]
Hi! 

I have just changed the namespace of "talk pin". It now compiles OK.

My TTL to RS485 converter needs it to "talk" to the network. Some other converters detects automatically. But mine's needs it.
If you use this kind of converter it does not matter to unuse the pin, but it have to be configured anyway.

I hope to test with my BMSs soon.!!

**Reply by virus100b** [reply]
Hi, @txubelaxu, 

First off all i want to take this opportunity to thank you for your work of developing the code for the new JK inverter BMS.

Regarding your code i am using Sleeper85 implementation of your code and i have a few question to ask. 

1. Using wired RS485 version  (ESP32-C3 super mini) and not BLE, can BMS switches be controlled somehow from Home Assistant?
2. A lot of sensors shows unknown.

![BMS02](https://github.com/txubelaxu/esphome-jk-bms/assets/64686560/a540900d-de3b-4f57-a45a-73957854fa9b)
![BMS02-1](https://github.com/txubelaxu/esphome-jk-bms/assets/64686560/aeb0cf29-31b4-4f5c-bc51-8938ed18fb34)
![BMS02-2](https://github.com/txubelaxu/esphome-jk-bms/assets/64686560/8ed9e190-1b6b-4fd8-b0d1-0a6b48be6bc2)

How can be that fixed? 

I must mention that in this moment the the CAN interface is not connected to the inverter. BMS is just sitting connected to the ESP and wifi. 

Thank you !

**Reply by txubelaxu** [reply]
Thank you very much for your appreciation. ;-) The truth is that the code is not as refined as I would like. I am quite a beginner in both programming and ESPHome. The code is neither finished nor optimized.

I am also using Sleeper85's code, and I see that the indicators you mentioned are showing correctly for me. Here is a printscreen.
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/156140720/aa6566a2-6987-42cf-a457-eafc9021dc68)

I think the ideal is to use "mode 2" (ESP32 acting as a REAL MASTER) for several reasons that, in turn, have a series of consequences:

It is the only way to retrieve 100% of the information from each and every node. The reason for this is that acting in "listening mode", i.e., having a BMS master (0x00), means acting as a pseudo-master taking advantage of communication silences; and also, it is not possible to obtain all the information from the BMS master.
In this mode, the ESP32 can issue commands to any of the BMS. I have not yet developed the part where ANY setting of any BMS on the network in client mode can be established.
The consequence of this mode of operation is that there will not be a BMS acting as a MASTER, and the ESP32 will have to act as an intermediary with the inverters (via CAN BUS).

**Reply by virus100b** [reply]
I definitely will use mode 2, but new i have just one BMS and is  sitting connected to the battery bank but not connected to the inverter. I will have another battery and inverter BMS soon and then with both connected in parallel maybe will work different.  Currently I modified the code and address dipswitch of BMS to 01 and i see some of entities "Unknown" and i am not able to change the state of control switches but i understand that they are just "read only" on rs485 wired version.

**Reply by Sleeper85** [reply]
> space

Thanks, I confirm that the compilation is now OK with `board_esp32-devkit-v1.yaml`.

**Comment by txubelaxu** [comment]
Yes, at this moment they are read only.

But the technology allow to change them. Only need development. :-))

**Comment by virus100b** [comment]
Some of the unknown entities was fixed after i have modified in yaml code address for BMS 1 from 00 to 01 and BMS 2 to address 02.  Before i had BMS1 set with address 00 and BMS 2 with address 01. The BMS had the dipswitches set to address 01, so the code was not recognizing any BMS with address 00. 
Now with this setting i have just a few entities with unknown state

![BMS001](https://github.com/txubelaxu/esphome-jk-bms/assets/64686560/38a0d106-3c25-46b5-a05f-b818d5f75eb7)

- For smartbms JK-BMS 1 operation status i don't know why is displaying that. Maybe because i don't have any current flowing trough the BMS
- The ones for power tube i think are wrong because of the name. You can see that the code is reading the values and display them as MOS temperatures protection and recovery and not as power tube.

**Comment by Sleeper85** [comment]
Hi guys,

I just updated my README for the solution with the new JK-BMS PB series and @txubelaxu component.

I don't really like the RS485 board used at the moment. A good RS485 board does not need a TALK PIN and is isolated.

So I added in the README the RS485 board that we use with the PVbrain2 PCB.

@txubelaxu or else, anyone want to test it? (I don't have a new JK-BMS).

[README JK-BMS-PB-CAN solution](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/documents/README/JK-BMS-PB-CAN_solution.md)
